File: docs/Sword2offer/CH2-编程题目-基础部分/ES6/combineStacks2Queue.js
let Stack = function () {
  let items = []
  let length = 0
  this.push = (value) => {
    items.push(value)
    length += 1
  }
  this.pop = () => {
    if (!this.isEmpty()) {
      let value = items.pop()
      length -= 1
      return value
    }
    return null
  }
  this.isEmpty = () => {
    return length === 0
  }
}

let Queue = function () {
  let stackIn = new Stack()
  let stackOut = new Stack()
  // 只有在stackout为空的时候，才进行将stackIn压入stackOut
  function makeStackOut () {
    if (stackOut.isEmpty()) {
      while (!stackIn.isEmpty()) {
        stackOut.push(stackIn.pop())
      }
    }
  }
  this.push = (value) => {
    stackIn.push(value)
  }
  this.shift = () => {
    makeStackOut()
    if (!stackOut.isEmpty()) {
      return stackOut.pop()
    }
    return null
  }
}

let queue = new Queue()
// 测试队列基本功能
queue.push('a')
queue.push('b')
queue.push('c')
console.log(queue.shift())
console.log(queue.shift())
// 测试stackOut不是空的压入情况
queue.push('d')
console.log(queue.shift())
console.log(queue.shift())

File: docs/Sword2offer/CH2-编程题目-基础部分/Python/combineStacks2Queue.py
class Stack:
  def __init__(self):
    self.__items = []
    self.__length = 0
  
  def append(self, value):
    self.__items.append(value)
    self.__length += 1
  
  def pop(self):
    if not self.is_empty():
      value = self.__items.pop()
      self.__length -= 1
      return value
    return None
  
  def is_empty(self):
    return self.__length == 0

  def test(self):
    return self.__items

class Queue:
  def __init__(self):
    self.__in = Stack()
    self.__out = Stack()
  
  def append(self, value):
    self.__in.append(value)
  
  def shift(self):
    self.__make_stackout()
    if not self.__out.is_empty():
      return self.__out.pop()
    return None

  def __make_stackout(self):
    if self.__out.is_empty():
      while not self.__in.is_empty():
        self.__out.append(self.__in.pop())

if __name__ == '__main__':
  queue = Queue()
  queue.append('a')
  queue.append('b')
  queue.append('c')
  print(queue.shift())
  print(queue.shift())
  queue.append('d')
  print(queue.shift())
  print(queue.shift())